Sprint 7 solution in file manager

IvanGriko opened the pull request:
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.

IvanGriko committed:
ФЗ 7

IvanGriko committed:
ФЗ 7.
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.

IvanGriko committed:
ФЗ 7.
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.

IvanGriko committed:
ФЗ 7.
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.

IvanGriko committed:
ФЗ 7.
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.

IvanGriko committed:
ФЗ 7.
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.

IvanGriko committed:
ФЗ 7.
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.

Yang-Pi commented on src/ru/yandex/tasktracker/exceptions/ManagerSaveException.java:
```diff
@@ -0,0 +1,10 @@
+package ru.yandex.tasktracker.exceptions;
+
+public class ManagerSaveException extends Exception {
```
чтобы не прописывать во всех методах throws ManagerSaveException, лучше унаследоваться от RuntimeException. По сути это исключение и есть непроверяемое исключение. Проверяемые исключения должны использоваться в тех случаях, когда у пользователей кода есть возможность восстановиться после ошибки и продолжить работу. Здесь же происходит завершение работы. (если интересно подробнее, можно почитать раздел item 71 Effective Java)

Yang-Pi commented on src/ru/yandex/tasktracker/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    public void save() throws ManagerSaveException {
+        if (file != null) {
```
лучше написать через guard-clause, то есть инвертировать проверку и выйти из метода. Тогда уровень вложенности уменьшится и последующий код будет красиво расположен на одном уровне с if

Yang-Pi commented on src/ru/yandex/tasktracker/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,278 @@
+
+    private static int taskCount;
+    private File file;
+    private final Map<Integer, Task> tasksMap = new HashMap<>();
```
у тебя уже есть мапы тасок в родительском классе, здесь они не нужны)

Yang-Pi commented on src/ru/yandex/tasktracker/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,278 @@
+                if (line.isEmpty()) {
+                    break;
+                } else if (task.getType().equals(TaskType.TASK)) {
+                    tasksMap.put(id, task);
```
добавлять таски лучше через методы интерфейса TaskManager - у тебя как раз создан объект backedTaskManager, через который ты сможешь добавить все задачи в соответствующие мапы

Yang-Pi commented on src/ru/yandex/tasktracker/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    @Override
+    public int getTaskCount() {
```
какой смысл в этом переопределении, если вызывается только метод родителя?)

Yang-Pi commented on src/ru/yandex/tasktracker/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,278 @@
+    }
+
+    @Override
+    public List<Epic> getEpics() throws ManagerSaveException {
```
эти переопределения тоже не нужны

Yang-Pi commented on src/ru/yandex/tasktracker/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,278 @@
+
+    private String taskToString(Object task) throws ManagerSaveException {
+        String string = "";
+        if (task.getClass().equals(Task.class)) {
```
давай воспользуемся ООП) можно переопределить метод toString() в классе Task, а также в его наследниках. Тогда все тело этого метода сведется к вызову task.toString()

Yang-Pi commented on src/ru/yandex/tasktracker/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,278 @@
+                    String name = str[2];
+                    String description = str[4];
+                    TaskStatus status = TaskStatus.valueOf(str[3]);
+                    task = new Task(name, description, id, status);
```
лучше ставить break в конце каждого case

Yang-Pi commented on src/ru/yandex/tasktracker/service/Managers.java:
```diff
@@ -1,5 +1,7 @@
+
 public class Managers {
 
     public static TaskManager getDefaultTaskManager() {
```
public static TaskManager getDefault() реализацию InMemoryTasksManager лучше заменить на FileBackedTasksManager

Yang-Pi commented on src/ru/yandex/tasktracker/service/Managers.java:
```diff
@@ -9,4 +11,8 @@ public static TaskManager getDefaultTaskManager() {
         return new InMemoryHistoryManager();
     }
+
+    public static FileBackedTaskManager getBackup(File file) {
```
отдельный метод не нужен

Yang-Pi reviewed (commented):
хорошая работа! оставил несколько замечаний, как можно улучшить реализацию. Если будут вопросы - пиши

IvanGriko committed:
ФЗ 7.1.
Добавить реализацию бекенда через запись данных в файл и восстановление из файла.
